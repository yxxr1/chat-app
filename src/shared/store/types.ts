import { CONNECTION_METHODS } from '@const/settings';

export type Chat = {
  id: string;
  name: string;
  messages: (Message | undefined)[];
  joinedCount: number | null;
};

export type Message = {
  id: string;
  text: string | null;
  fromId: string;
  fromName: string;
  date: number;
  service?: number;
  index: number;
};

export type UserSettings = {
  connectionMethod: (typeof CONNECTION_METHODS)[keyof typeof CONNECTION_METHODS];
  theme: 'light' | 'dark';
};

export type User = {
  id: string;
  name: string;
  settings: UserSettings;
};

export interface State {
  isLoading: boolean;
  user: User | null;
  allChats: { [chatId: Chat['id']]: Chat };
  joinedChatsIds: Chat['id'][];
  subscribedChatsIds: Chat['id'][];
  currentChatId: Chat['id'] | null;
}
