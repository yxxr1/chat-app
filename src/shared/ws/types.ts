import { Message, Chat } from '@store/types';

type WSMessage<Type, Payload> = {
  type: Type;
  payload: Payload;
};

export type WatchChats = WSMessage<
  'WATCH_CHATS',
  {
    newChats: Chat[];
    deletedChatsIds: Chat['id'][];
    updatedChats: Chat[];
  }
>;

export type ChatSubscribe = WSMessage<
  'SUBSCRIBED_CHAT',
  {
    chatId: Chat['id'];
    messages: Message[];
  }
>;

export type WSMessageIncoming = ChatSubscribe | WatchChats;

export type WSMessageOutgoing =
  | WSMessage<
      'PUBLISH_MESSAGE',
      {
        chatId: Chat['id'];
        message: Message['text'];
      }
    >
  | WSMessage<
      'SUBSCRIBE_CHAT',
      {
        chatId: Chat['id'];
        lastMessageId: Message['id'] | null;
      }
    >;
