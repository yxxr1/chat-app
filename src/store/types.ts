import {
  addMessages,
  joinChat,
  quitChat,
  setChatList,
  setCurrentChat,
  setUser,
  addChats,
  deleteChats,
  addSubscribedChats,
} from '@actions/sync';
import { CONNECTION_METHODS } from '@const/settings';

export type Chat = {
  id: string;
  name: string;
  messages: (Message | undefined)[];
};

export type Message = {
  id: string;
  text: string | null;
  fromId: string;
  fromName: string;
  date: number;
  service?: number;
  index: number;
};

export type UserSettings = {
  connectionMethod: (typeof CONNECTION_METHODS)[keyof typeof CONNECTION_METHODS];
};

export type User = {
  id: string;
  name: string;
  settings: UserSettings;
};

export type State = {
  user: User | null;
  isLoading: boolean;
  allChats: { [chatId: Chat['id']]: Chat };
  joinedChatsIds: Chat['id'][];
  subscribedChatsIds: Chat['id'][];
  currentChatId: Chat['id'] | null;
};

export type Action =
  | ReturnType<typeof addMessages>
  | ReturnType<typeof joinChat>
  | ReturnType<typeof quitChat>
  | ReturnType<typeof setChatList>
  | ReturnType<typeof setCurrentChat>
  | ReturnType<typeof setUser>
  | ReturnType<typeof addChats>
  | ReturnType<typeof deleteChats>
  | ReturnType<typeof addSubscribedChats>;
