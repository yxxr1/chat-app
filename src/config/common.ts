export const COMMON_CONFIG = {
  API_URL: process.env.API_URL || 'http://localhost:8080',
  WS_URL: process.env.WS_URL || 'ws://localhost:8080/ws',
};
