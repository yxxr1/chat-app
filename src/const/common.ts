export const MESSAGE_SERVICE_TYPES = {
  CHAT_JOINED: 1,
  CHAT_LEFT: 2,
} as const;
