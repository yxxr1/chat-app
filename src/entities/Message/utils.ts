import moment from 'moment';
import { Message as MessageType } from '@store/types';

export const getUserTitle = (message: MessageType) => `${message.fromName}(${message.fromId.substr(message.fromId.length - 4)})`;

export const formatDate = (date: number) => moment(date).calendar();
