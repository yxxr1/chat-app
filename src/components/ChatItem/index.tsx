import React from 'react';
import { Message as MessageType } from '@store/types';
import { getServiceMessage } from '@utils/common';
import { Chat, Title, Message } from './styled';

type Props = {
  name: string;
  lastMessage: MessageType | null;
  isCurrent: boolean;
  onClick: () => void;
};

export const ChatItem: React.FC<Props> = ({ name, lastMessage, isCurrent, onClick }) => {
  return (
    <Chat isCurrent={isCurrent} onClick={onClick}>
      <Title>{name}</Title>
      {lastMessage && (
        <Message>
          {lastMessage.service ? (
            <i>{getServiceMessage(lastMessage)}</i>
          ) : (
            <>
              <i>{lastMessage.fromName}:</i>
              {` ${lastMessage.text}`}
            </>
          )}
        </Message>
      )}
    </Chat>
  );
};
