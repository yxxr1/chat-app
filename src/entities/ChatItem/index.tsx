import React, { memo, useMemo } from 'react';
import { Message as MessageType, Chat as ChatType } from '@store/types';
import { getServiceMessage } from '@utils/common';
import { Chat, Title, Message } from './styled';

type Props = {
  chat: ChatType;
  isCurrent: boolean;
  onClick: (chatId: ChatType['id']) => void;
};

const _ChatItem: React.FC<Props> = ({ chat, isCurrent, onClick }) => {
  const lastMessage = useMemo<MessageType | null>(() => chat.messages.slice(-1)[0] ?? null, [chat.messages]);

  return (
    <Chat data-testid="chatItem__container" isCurrent={isCurrent} onClick={() => onClick(chat.id)}>
      <Title>{chat.name}</Title>
      {lastMessage && (
        <Message>
          {lastMessage.service ? (
            <i>{getServiceMessage(lastMessage)}</i>
          ) : (
            <>
              <i>{lastMessage.fromName}:</i>
              {` ${lastMessage.text}`}
            </>
          )}
        </Message>
      )}
    </Chat>
  );
};

export const ChatItem = memo(_ChatItem);
