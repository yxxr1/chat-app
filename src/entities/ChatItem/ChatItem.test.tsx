import React from 'react';
import { cleanup, fireEvent, screen } from '@testing-library/react';
import { ChatItem } from '@entities/ChatItem';
import { Chat, Message } from '@store/types';
import { renderComponent } from '@test/utils';
import { THEMES } from '@styles/theme';
import { MESSAGE_SERVICE_TYPES } from '@const/common';

const chat: Chat = {
  id: '1',
  name: 'test chat',
  messages: [
    {
      id: '1',
      text: 'test message',
      fromId: '1',
      fromName: 'testUser',
      date: new Date().valueOf(),
      index: 0,
    },
  ],
  joinedCount: 2,
};

describe('ChatItem renders', () => {
  afterEach(() => {
    cleanup();
  });

  test('renders content', () => {
    renderComponent(<ChatItem chat={chat} isCurrent={false} onClick={() => {}} />);
    expect(screen.getByTestId('chatItem__container')).toMatchSnapshot();
  });

  test('renders content no messages', () => {
    renderComponent(<ChatItem chat={{ ...chat, messages: [] }} isCurrent={false} onClick={() => {}} />);
    expect(screen.getByTestId('chatItem__container')).toMatchSnapshot();
  });

  test('renders content service message', () => {
    renderComponent(
      <ChatItem
        chat={{ ...chat, messages: [{ ...chat.messages[0], service: MESSAGE_SERVICE_TYPES.CHAT_JOINED } as Message] }}
        isCurrent={false}
        onClick={() => {}}
      />,
    );
    expect(screen.getByTestId('chatItem__container')).toMatchSnapshot();
  });

  test('render isCurrent=false, theme light', () => {
    renderComponent(<ChatItem chat={chat} isCurrent={false} onClick={() => {}} />, 'light');
    expect(screen.getByTestId('chatItem__container')).toHaveStyle({
      'background-color': THEMES.light['background-alt'],
      border: `solid 1px ${THEMES.light.border}`,
    });
  });

  test('render isCurrent=true, theme light', () => {
    renderComponent(<ChatItem chat={chat} isCurrent={true} onClick={() => {}} />, 'light');
    expect(screen.getByTestId('chatItem__container')).toHaveStyle({
      'background-color': THEMES.light['background'],
      border: `solid 1px ${THEMES.light.primary}`,
    });
  });

  test('render isCurrent=false, theme dark', () => {
    renderComponent(<ChatItem chat={chat} isCurrent={false} onClick={() => {}} />, 'dark');
    expect(screen.getByTestId('chatItem__container')).toHaveStyle({
      'background-color': THEMES.dark['background-alt'],
      border: `solid 1px ${THEMES.dark.border}`,
    });
  });

  test('render isCurrent=true, theme dark', () => {
    renderComponent(<ChatItem chat={chat} isCurrent={true} onClick={() => {}} />, 'dark');
    expect(screen.getByTestId('chatItem__container')).toHaveStyle({
      'background-color': THEMES.dark['background'],
      border: `solid 1px ${THEMES.dark.primary}`,
    });
  });

  test('onClick', () => {
    const handler = jest.fn();
    renderComponent(<ChatItem chat={chat} isCurrent={false} onClick={handler} />);
    fireEvent.click(screen.getByTestId('chatItem__container'));
    fireEvent.click(screen.getByTestId('chatItem__container'));
    expect(handler).toHaveBeenCalledTimes(2);
    expect(handler).toHaveBeenCalledWith(chat.id);
  });
});
