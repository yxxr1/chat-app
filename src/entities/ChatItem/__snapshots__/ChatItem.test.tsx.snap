// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ChatItem renders renders content 1`] = `
<div
  class="sc-aXZVg rugVN"
  data-testid="chatItem__container"
>
  <div
    class="sc-gEvEer eTINNo"
  >
    test chat
  </div>
  <div
    class="sc-eqUAAy MWRcI"
  >
    <i>
      testUser
      :
    </i>
     test message
  </div>
</div>
`;

exports[`ChatItem renders renders content no messages 1`] = `
<div
  class="sc-aXZVg rugVN"
  data-testid="chatItem__container"
>
  <div
    class="sc-gEvEer eTINNo"
  >
    test chat
  </div>
</div>
`;

exports[`ChatItem renders renders content service message 1`] = `
<div
  class="sc-aXZVg rugVN"
  data-testid="chatItem__container"
>
  <div
    class="sc-gEvEer eTINNo"
  >
    test chat
  </div>
  <div
    class="sc-eqUAAy MWRcI"
  >
    <i>
      testUser joined chat
    </i>
  </div>
</div>
`;
